src/analytics/private/segments/user-segments-workflow.ts

```diff
@@ -0,0 +1,217 @@
+    const executeSegmentQueryFunc = this.constructNodejsFunction('execute-segment-query', [
+      new PolicyStatement({
+        actions: [
+          'dynamodb:GetItem',
+          'dynamodb:UpdateItem',
+        ],
+        resources: [props.clickstreamMetadataDdbArn],
+      }),
+    ], {
+      REDSHIFT_MODE: props.serverlessRedshift ? REDSHIFT_MODE.SERVERLESS : REDSHIFT_MODE.PROVISIONED,
+      REDSHIFT_SERVERLESS_WORKGROUP_NAME: props.serverlessRedshift?.workgroupName ?? '',
+      REDSHIFT_CLUSTER_IDENTIFIER: props.provisionedRedshift?.clusterIdentifier ?? '',
+      REDSHIFT_DATABASE: props.databaseName,
+      REDSHIFT_DB_USER: props.provisionedRedshift?.dbUser ?? '',
+      REDSHIFT_DATA_API_ROLE: props.dataAPIRole.roleArn,
+      CLICKSTREAM_METADATA_DDB_ARN: props.clickstreamMetadataDdbArn,
+    });
+    const executeSegmentQueryTask = new LambdaInvoke(this, 'WorkflowTask-ExecuteSegmentQuery', {
+      lambdaFunction: executeSegmentQueryFunc,
+      outputPath: '$.Payload',
+    });
+    props.dataAPIRole.grantAssumeRole(executeSegmentQueryFunc.grantPrincipal);
+
+    // Define task for checking segment job status
+    const segmentJobStatusFunc = this.constructNodejsFunction('segment-job-status', [
+      new PolicyStatement({
+        actions: ['dynamodb:UpdateItem'],
+        resources: [props.clickstreamMetadataDdbArn],
+      }),
+    ], {
+      REDSHIFT_DATA_API_ROLE: props.dataAPIRole.roleArn,
+      CLICKSTREAM_METADATA_DDB_ARN: props.clickstreamMetadataDdbArn,
+      SEGMENTS_S3_PREFIX: props.segmentsS3Prefix,
+    });
+    const segmentJobStatusTask = new LambdaInvoke(this, 'WorkflowTask-SegmentJobStatus', {
+      lambdaFunction: segmentJobStatusFunc,
+      outputPath: '$.Payload',
+    });
+    props.dataAPIRole.grantAssumeRole(segmentJobStatusFunc.grantPrincipal);
+    Bucket.fromBucketName(this, 'PipelineS3Bucket', props.pipelineS3Bucket).grantRead(segmentJobStatusFunc, `${props.segmentsS3Prefix}*`);
+
+    // Define Succeed and Fail end state
+    const succeedState = new Succeed(this, 'WorkflowEndState-Succeed');
+    const failState = new Fail(this, 'WorkflowEndState-Fail');
+
+    // Define choice state for workflow initialization
+    // If refresh schedule has expired, go to end state
+    const segmentJobInitChoice = new Choice(this, 'WorkflowChoice-SegmentJobInit')
+      .when(Condition.booleanEquals('$.scheduleIsExpired', true), succeedState)
+      .otherwise(stateMachineStatusTask);
+
+    // Connect segment job init task to the choice
+    segmentJobInitTask.next(segmentJobInitChoice);
+
+    // Define choice state for checking state machine status
+    // If state machine is idle, execute segment query. Otherwise, wait for 1 min and check again
+    const stateMachineStatusChoice = new Choice(this, 'WorkflowChoice-StateMachineStatus')
+      .when(Condition.stringEquals('$.stateMachineStatus', 'IDLE'), executeSegmentQueryTask)
+      .otherwise(new Wait(this, 'WorkflowWait-StateMachineStatus', {
```

Comment:
Let's have the backoff support introduced in #943.

Reply:
done